stats/redis.go

```diff
@@ -117,15 +136,86 @@ func (l *redisListener) HandleEvent(event events.Event) {
 	namespace := statsNamespace(key, event.Namespace())
 	bucket := event.BucketName()
 
-	var incr *redis.FloatCmd
-	_, err := l.client.Pipelined(func(pipe *redis.Pipeline) error {
-		incr = pipe.ZIncrBy(namespace, float64(numTokens), bucket)
-		pipe.ExpireAt(namespace, nearestHour())
-		return nil
-	})
+	l.queueStatsUpdate(namespace, numTokens, bucket)
+}
+
+// queueStatsUpdate queues a statsUpdate to be sent to redis via the batcher
+func (l *redisListener) queueStatsUpdate(namespace string, numTokens int64, bucket string) {
+	l.statsUpdatesLock.Lock()
+
+	l.pipe.ZIncrBy(namespace, float64(numTokens), bucket)
+	l.pipe.ExpireAt(namespace, nearestHour())
+
+	l.queuedUpdates++
+
+	l.statsUpdatesLock.Unlock()
+
+	select {
+	case l.notifyBatcher <- struct{}{}:
+		// Done
+	default:
+		// There's already a pending notification
+	}
+}
+
+// batcher aggregates stats updates. Any stats that are aggregated are guaranteed
+// to be sent within batchSubmitInterval, or once batchSize stats updates are queued,
+// whichever happens first.
+func (l *redisListener) batcher() {
+	timeout := time.After(l.batchDeadline)
+
+	for {
+		select {
+		case <-timeout:
```

Comment:
You can call `case<-time.After(l.batchDeadline)` here instead of reassigning the variable in a couple of places below.

Reply:
I actually can't, it's a subtle bug.  If you do `case <- time.After`, it can actually take roughly `l.batchDeadline * l.batchSize` to send the oldest stats update, since it would be starting the timeout anew every time it receives a new stats update.

If that's the behaviour we want I could just do that, but I was trying to make sure every stats update would be submitted within roughly `l.batchDeadline`.

Reply:
Even when you're submitting the batch in a separate, non-blocking goroutine?

I defer to your judgment.